### PR TITLE
Fixing compatibility problems with PyTorch 1.3.0 in ESPnet (v0.5.3)

### PR DESCRIPTION
**Pull request by Emrys365**
Hi, I am trying to make ESPnet support the latest PyTorch version. Here are some fixed problems.

1. Inconsistent ctc_loss API
    + Since `warp_ctc` currently does not work with PyTorch 1.3, I am using `builtin` CTC implementation in PyTorch. But I met the same problem as [issue#22234](https://github.com/pytorch/pytorch/issues/22234).
    + In this pull request, I fix this issue by using the deterministic cuDNN implementation of CTC loss. This is done by wrapping the computation of CTC loss with:
      ```python
      with torch.backends.cudnn.flags(deterministic=True):
      ```
2. Warnings of using `torch.uint8` for `masked_fill` and indexing
    + After switching to PyTorch 1.3, tons of warnings of the following two types appeared during training:
      ```
      /pytorch/aten/src/ATen/native/cuda/LegacyDefinitions.cpp:19: UserWarning: masked_fill_ received a mask with dtype torch.uint8, this behavior is now deprecated,please use a mask with dtype torch.bool instead.
      /pytorch/aten/src/ATen/native/IndexingUtils.h:20: UserWarning: indexing with dtype torch.uint8 is now deprecated, please use a dtype torch.bool instead.
      ```
    + I managed to get rid of the first warning by replacing all `torch.uint8` with `torch.bool` when using `masked_fill()` or yielding masks.
    + For the second warning, however, I didn't find the code responsible for that. I will figure it out maybe later.

-------
### Update
I found the cause of the warning `indexing with dtype torch.uint8 is now deprecated`. It comes from [torch_complex.functional.trace()](https://github.com/kamo-naoyuki/pytorch_complex/blob/13f9bba1cc23df4ed859cb13cec6a7a9873bbd23/torch_complex/functional.py#L126), where the code explicitly uses `torch.uint8` for indexing.

### CODE REVIEW

**Comment by Emrys365** [comment]
There are still some unsolved issues with PyTorch 1.3.0. I will list them below.
I would appreciate it if anyone could provide some help!

1. When running the WSJ recipe with 1 GPU, everything  is fine.
While running the WSJ recipe with 2 GPUs, the following error occurred:
```python
[J    total [###...............................................]  6.57%
this epoch [#################################################.] 98.58%
      1800 iter, 0 epoch / 15 epochs
   0.80658 iters/sec. Estimated time to finish: 8:48:46.519785.
[4A[Jxxx/espnet-v0.5.3/espnet/nets/pytorch_backend/e2e_asr.py:338: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  hs_pad, torch.tensor(hlens), lpz,
Exception in main training loop: Caught RuntimeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
    output = module(*input, **kwargs)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "xxx/espnet-v0.5.3/espnet/nets/pytorch_backend/e2e_asr.py", line 287, in forward
    hs_pad, hlens, _ = self.enc(hs_pad, hlens)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "xxx/espnet-v0.5.3/espnet/nets/pytorch_backend/rnn/encoders.py", line 256, in forward
    xs_pad, ilens, states = module(xs_pad, ilens, prev_state=prev_state)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "xxx/espnet-v0.5.3/espnet/nets/pytorch_backend/rnn/encoders.py", line 65, in forward
    rnn.flatten_parameters()
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/rnn.py", line 120, in flatten_parameters
    self.batch_first, bool(self.bidirectional))
RuntimeError: set_storage is not allowed on a Tensor created from .data or .detach().
If your intent is to change the metadata of a Tensor (such as sizes / strides / storage / storage_offset)
without autograd tracking the change, remove the .data / .detach() call and wrap the change in a `with torch.no_grad():` block.
For example, change:
    x.data.set_(y)
to:
    with torch.no_grad():
        x.set_(y)

Traceback (most recent call last):
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/chainer/training/trainer.py", line 319, in run
    entry.extension(self)
  File "xxx/espnet-v0.5.3/espnet/utils/training/evaluator.py", line 10, in __call__
    ret = super().__call__(trainer)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/chainer/training/extensions/evaluator.py", line 161, in __call__
    result = self.evaluate()
  File "xxx/espnet-v0.5.3/espnet/asr/pytorch_backend/asr.py", line 127, in evaluate
    data_parallel(self.model, x, range(self.ngpu))
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 207, in data_parallel
    outputs = parallel_apply(replicas, inputs, module_kwargs, used_device_ids)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 85, in parallel_apply
    output.reraise()
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/_utils.py", line 385, in reraise
    raise self.exc_type(msg)
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "xxx/espnet-v0.5.3/egs/wsj/asr1/../../../espnet/bin/asr_train.py", line 368, in <module>
    main(sys.argv[1:])
  File "xxx/espnet-v0.5.3/egs/wsj/asr1/../../../espnet/bin/asr_train.py", line 349, in main
    train(args)
  File "xxx/espnet-v0.5.3/espnet/asr/pytorch_backend/asr.py", line 623, in train
    trainer.run()
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/chainer/training/trainer.py", line 349, in run
    six.reraise(*exc_info)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/six.py", line 693, in reraise
    raise value
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/chainer/training/trainer.py", line 319, in run
    entry.extension(self)
  File "xxx/espnet-v0.5.3/espnet/utils/training/evaluator.py", line 10, in __call__
    ret = super().__call__(trainer)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/chainer/training/extensions/evaluator.py", line 161, in __call__
    result = self.evaluate()
  File "xxx/espnet-v0.5.3/espnet/asr/pytorch_backend/asr.py", line 127, in evaluate
    data_parallel(self.model, x, range(self.ngpu))
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 207, in data_parallel
    outputs = parallel_apply(replicas, inputs, module_kwargs, used_device_ids)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 85, in parallel_apply
    output.reraise()
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/_utils.py", line 385, in reraise
    raise self.exc_type(msg)
RuntimeError: Caught RuntimeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
    output = module(*input, **kwargs)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "xxx/espnet-v0.5.3/espnet/nets/pytorch_backend/e2e_asr.py", line 287, in forward
    hs_pad, hlens, _ = self.enc(hs_pad, hlens)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "xxx/espnet-v0.5.3/espnet/nets/pytorch_backend/rnn/encoders.py", line 256, in forward
    xs_pad, ilens, states = module(xs_pad, ilens, prev_state=prev_state)
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "xxx/espnet-v0.5.3/espnet/nets/pytorch_backend/rnn/encoders.py", line 65, in forward
    rnn.flatten_parameters()
  File "xxx/espnet-v0.5.3/tools/venv/lib/python3.7/site-packages/torch/nn/modules/rnn.py", line 120, in flatten_parameters
    self.batch_first, bool(self.bidirectional))
RuntimeError: set_storage is not allowed on a Tensor created from .data or .detach().
If your intent is to change the metadata of a Tensor (such as sizes / strides / storage / storage_offset)
without autograd tracking the change, remove the .data / .detach() call and wrap the change in a `with torch.no_grad():` block.
For example, change:
    x.data.set_(y)
to:
    with torch.no_grad():
        x.set_(y)
```

--------
#### to be  continued

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (sw005320) [inline .vscode/settings.json]: Can you remove this file?

Comment (Emrys365) [reply .vscode/settings.json]: Oh, sorry! I should have removed it.

**Comment by sw005320** [comment]
Can you fix some of ci errors?

**Comment by Emrys365** [comment]
> Can you fix some of ci errors?

Yes. It is now all clear.

**Comment by Emrys365** [comment]
> LGTM
> The remaining issue is the same as [pytorch/pytorch#21108](https://github.com/pytorch/pytorch/issues/21108)?

I think so. The error occurred when doing evaluation after training for the first epoch. It is the same case as [pytorch/pytorch#21108].

**Comment by sw005320** [comment]
> ### Update
> I found the cause of the warning `indexing with dtype torch.uint8 is now deprecated`. It comes from [torch_complex.functional.trace()](https://github.com/kamo-naoyuki/pytorch_complex/blob/13f9bba1cc23df4ed859cb13cec6a7a9873bbd23/torch_complex/functional.py#L126), where the code explicitly uses `torch.uint8` for indexing.

Very cool!
@kamo-naoyuki, could you check https://github.com/kamo-naoyuki/pytorch_complex/pull/3 ?

**Comment by sw005320** [comment]
> > LGTM
> > The remaining issue is the same as [pytorch/pytorch#21108](https://github.com/pytorch/pytorch/issues/21108)?
> 
> I think so. The error occurred when doing evaluation after training for the first epoch. It is the same case as [[pytorch/pytorch#21108](https://github.com/pytorch/pytorch/issues/21108)].

Then, can you add some `assert` to avoid the above error in the pytorch 1.3 multiple GPU case? Then, I'll merge it and wait for pytorch's bugfix.

**Comment by Emrys365** [comment]
> > > LGTM
> > > The remaining issue is the same as [pytorch/pytorch#21108](https://github.com/pytorch/pytorch/issues/21108)?
> > 
> > 
> > I think so. The error occurred when doing evaluation after training for the first epoch. It is the same case as [[pytorch/pytorch#21108](https://github.com/pytorch/pytorch/issues/21108)].
> 
> Then, can you add some `assert` to avoid the above error in the pytorch 1.3 multiple GPU case? Then, I'll merge it and wait for pytorch's bugfix.

Yes. I modified `espnet/bin/asr_train.py` to block multi-GPU training with PyTorch 1.2+.

**Comment by sw005320** [comment]
This is a great job.
Thanks @Emrys365 !

**Comment by Emrys365** [comment]
I think we can remove [these lines](https://github.com/espnet/espnet/blob/master/espnet/bin/asr_train.py#L576-L580) after PyTorch fixes the problem I mentioned here: https://github.com/pytorch/pytorch/pull/21197#issuecomment-661616220

I have tested manually by replacing `replicas = replicate(module, used_device_ids)` in [torch/nn/parallel/data_parallel.py#L217](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/data_parallel.py#L217) by `replicas = replicate(module, used_device_ids, not torch.is_grad_enabled())`, and it works now!

**Pull request merged**